clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -6879,3 +6879,1781 @@ void SemaSYCL::handleKernelAttr(Decl *D, const ParsedAttr &AL) {
+  if (const auto *DeclAttr = D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>()) {
+    if (areInvalidWorkGroupSizeAttrs(DeclAttr->getValue(), XDim, YDim, ZDim)) {
+      Diag(CI.getLoc(), diag::err_sycl_x_y_z_arguments_must_be_one)
+          << CI << DeclAttr;
+    }
+  }
+
+  // If the 'max_work_group_size' attribute is specified on a declaration along
+  // with 'reqd_work_group_size' attribute, check to see if values of
+  // 'reqd_work_group_size' attribute arguments are equal to or less than values
+  // of 'max_work_group_size' attribute arguments.
+  //
+  // We emit diagnostic if values of 'reqd_work_group_size' attribute arguments
+  // are greater than values of 'max_work_group_size' attribute arguments.
+  if (const auto *DeclAttr = D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
+    if (checkMaxAllowedWorkGroupSize(XDim, YDim, ZDim, DeclAttr->getXDim(),
+                                     DeclAttr->getYDim(),
+                                     DeclAttr->getZDim())) {
+      Diag(CI.getLoc(), diag::err_conflicting_sycl_function_attributes)
+          << CI << DeclAttr;
+      Diag(DeclAttr->getLoc(), diag::note_conflicting_attribute);
+      return;
+    }
+  }
+
+  // If the 'reqd_work_group_size' attribute is specified on a declaration
+  // along with 'num_simd_work_items' attribute, the required work group size
+  // specified by 'num_simd_work_items' attribute must evenly divide the index
+  // that increments fastest in the 'reqd_work_group_size' attribute.
+  if (const auto *DeclAttr = D->getAttr<SYCLIntelNumSimdWorkItemsAttr>()) {
+    if (checkWorkGroupSize(DeclAttr->getValue(), XDim, YDim, ZDim)) {
+      Diag(DeclAttr->getLoc(), diag::err_sycl_num_kernel_wrong_reqd_wg_size)
+          << DeclAttr << CI;
+      Diag(CI.getLoc(), diag::note_conflicting_attribute);
+      return;
+    }
+  }
+
+  // If the attribute was already applied with different arguments, then
+  // diagnose the second attribute as a duplicate and don't add it.
+  if (const auto *Existing = D->getAttr<SYCLReqdWorkGroupSizeAttr>()) {
+    // If any of the results are known to be different, we can diagnose at this
+    // point and drop the attribute.
+    if (anyWorkGroupSizesDiffer(XDim, YDim, ZDim, Existing->getXDim(),
+                                Existing->getYDim(), Existing->getZDim())) {
+      Diag(CI.getLoc(), diag::err_duplicate_attribute) << CI;
+      Diag(Existing->getLoc(), diag::note_previous_attribute);
+      return;
+    }
+
+    // If all of the results are known to be the same, we can silently drop the
+    // attribute. Otherwise, we have to add the attribute and resolve its
+    // differences later.
+    if (allWorkGroupSizesSame(XDim, YDim, ZDim, Existing->getXDim(),
+                              Existing->getYDim(), Existing->getZDim()))
+      return;
+  }
+
+  ASTContext &Context = getASTContext();
+  D->addAttr(::new (Context)
+                 SYCLReqdWorkGroupSizeAttr(Context, CI, XDim, YDim, ZDim));
+}
```

Comment:
Overall, I think this is a great improvement over the current state, but I have concerns about the SemaSYCL.cpp size.

Can we move SYCL attributes handling logic to a separate source file (e.g. SemaSYCLAttrDecl.cpp or something similar)?
I think SemaSYCL.cpp already crossed the threshold for maintainable source file size. Adding another 2000 lines to the current 6500 lines makes SYCL specific Sema code harder to navigate.

Reply:
I've moved the new code to `SemaSYCLAttrDecl.cpp`

Reply:
I've named it `SemaSYCLDeclAttr.cpp` to better reflect `SemaDeclAttr.cpp`